src/path-graph.js

```diff
@@ -1033,10 +1138,181 @@ export class Graph extends Component {
 Graph = connect(
   (state) => ({
+    sourceNode: state.sourceNode,
+    targetNode: state.targetNode,
     pathQueries: state.pathQueries,
     hetioStyles: state.hetioStyles
   }),
   null,
   null,
   { forwardRef: true }
 )(Graph);
+
+// selected node info component
+class SelectedNodeInfo extends Component {
+  // display row entries
+  rows() {
+    // helper text when user hovers over given field
+    let tooltipText = {};
+    if (this.props.hetioDefinitions.properties) {
+      tooltipText = {
+        ...tooltipText,
+        ...this.props.hetioDefinitions.properties.common,
+        ...this.props.hetioDefinitions.properties.nodes
+      };
+    }
+    tooltipText = { ...tooltipText, ...this.props.hetmechDefinitions };
+    if (tooltipText['id'])
+      tooltipText['neo4j_id'] = tooltipText['id'];
+
+    // get primary fields from top level of node
+    let primaryFields = ['metanode', 'neo4j_id'];
+    // get first/second column text (key/value) for each field
+    primaryFields = primaryFields.map((field) => ({
+      firstCol: field,
+      secondCol: this.props.node[field],
+      tooltipText: tooltipText[field]
+    }));
+
+    // get 'extra fields' from node 'data' field
+    let extraFields = Object.keys(this.props.node.data);
+    // get first/second column text (key/value) for each field
+    extraFields = extraFields.map((field) => ({
+      firstCol: field,
+      secondCol: this.props.node.data[field],
+      tooltipText: tooltipText[field]
+    }));
+
+    // combine primary and extra fields
+    let fields = primaryFields.concat(extraFields);
+
+    // display fields in custom order
+    const order = [
+      'name',
+      'metanode',
+      'source',
+      'url',
+      'description',
+      'identifier',
+      'neo4j_id'
+    ];
+    fields = sortCustom(fields, order, 'firstCol');
+
+    // make columns from fields
+    const cols = fields.map((field, index) => {
+      return (
+        <React.Fragment key={index}>
+          <td className='col_s small light'>
+            <Tooltip text={field.tooltipText}>{field.firstCol}</Tooltip>
+          </td>
+          <td className='small'>{field.secondCol}</td>
+        </React.Fragment>
+      );
+    });
+
+    // make rows in groups of two
+    const rows = new Array(Math.ceil(cols.length / 2))
+      .fill()
+      .map(() => cols.splice(0, 2))
+      .map((col, index) => <tr key={index}>{col}</tr>);
+
+    return rows;
+  }
+
+  // display component
+  render() {
+    return (
+      <>
+        <div className='graph_info_header small light'>Selected Node</div>
+        <table id='graph_info_table'>
+          <tbody>{this.rows()}</tbody>
+        </table>
+      </>
+    );
+  }
+}
+// connect component to global state
+SelectedNodeInfo = connect((state) => ({
+  hetioDefinitions: state.hetioDefinitions,
+  hetmechDefinitions: state.hetmechDefinitions
+}))(SelectedNodeInfo);
+
+// selected edge info component
+class SelectedEdgeInfo extends Component {
```

Comment:
I see some code duplication in this class and `SelectedNodeInfo` class. How hard is it going to be to remove the duplication (such as defining a abstract class that will be inherited by `SelectedNodeInfo` and `SelectedEdgeInfo`)?

Reply:
Incorporated.

Reply:
Thanks.